src/units.jl

```diff
@@ -0,0 +1,106 @@
+module Units
+
+import ..DEFAULT_DIM_TYPE
+import ..DEFAULT_VALUE_TYPE
+import ..Quantity
+
+@assert DEFAULT_VALUE_TYPE == Float64 "`units.jl` must be updated to support a different default value type."
+
+macro add_prefixes(base_unit, prefixes)
+    @assert prefixes.head == :tuple
+    expr = _add_prefixes(base_unit, prefixes.args)
+    return expr |> esc
+end
+
+function _add_prefixes(base_unit::Symbol, prefixes)
+    all_prefixes = (
+        f=1e-15, p=1e-12, n=1e-9, μ=1e-6, u=1e-6, m=1e-3, c=1e-2,
+        k=1e3, M=1e6, G=1e9, T=1e12, P=1e15
+    )
+    expr = Expr(:block)
+    for (prefix, value) in zip(keys(all_prefixes), values(all_prefixes))
+        prefix in prefixes || continue
+        new_unit = Symbol(prefix, base_unit)
+        push!(expr.args, :(const $new_unit = $value * $base_unit))
+    end
+    return expr
+end
+
+# SI base units
+const m = Quantity(1.0, length=1)
```

Comment:
IMO  single leter names here are bad. I would go with `meter` `gram` etc.

Reply:
This is just for use in `uparse`, so you can write, e.g., `10u"km/h"`. This file is wrapped in a `module Units ... end` so none of these units are actually exported.

Reply:
Should these be tied to Float64? Using a smaller number type would allow e.g. `2 * Quantity(true, length=1)` to come out as an integer.

Reply:
> Should these be tied to Float64? Using a smaller number type would allow e.g. `2 * Quantity(true, length=1)` to come out as an integer.

This is fair, but (1) it would mean the units would have different types (e.g., `fm` would have a different type than `m`). And (2) I think the type of people who worry about 8-bit vs 64-bit types are likely to figure out they can use `convert(Quantity{Bool}, u"m/s")` when needed. Whereas the lay user will be left confused when they type out
```julia
julia> u"m/s"
true m s⁻¹

julia> u"km/s"
1000.0 m s⁻¹
```

Reply:
Oh, btw, your example is true regardless:

```julia
julia> typeof(2 * Quantity(true, length=1))
Quantity{Int64, DynamicQuantities.FixedRational{Int32, 25200}}
```

the types of *units* (given by `uparse` and `@u_str`) are Float64. But a `Quantity()` is whatever type you pass.

Reply:
Yea I know, sorry my point was roughly to wonder whether `2m` should give something that's an integer, not a Float64, and perhaps that `2f0 m` should be Float32. Maybe Bool is too exotic but `1` would do this.

Reply:
I guess we could make some `WeakFloat64` that stores a `Float64` but eagerly converts to the user’s type upon being used in any calculation, so long as there’s no rounding error?

But might introduce some type instability.

Reply:
Not sure I follow, how would that differ from `1::Int`? This is type-stable.

For numbers < 1, you could use Rational, or your FixedRational. Besides not promoting small types, this has the advantage of not reducing the accuracy of large ones. `(1e-3) * big(1.0)` is only Float64 accurate, while `(1//1000) * big(1.0) == big"0.001"`

Reply:
e.g., femtometers (=1e-15 m)

Reply:
Is this better @mcabbott?

```julia
    z = u"yr"
    @test typeof(z) == Quantity{Rational{Int64},DEFAULT_DIM_TYPE}

    z2 = 1.0u"yr"
    @test typeof(z2) == Quantity{Float64,DEFAULT_DIM_TYPE}

    z3 = 1f0u"yr"
    @test typeof(z3) == Quantity{Float32,DEFAULT_DIM_TYPE}
```

So now the default type of **all** units is Rational{Int}. Thus, promotion to floats works as you might expect.

Reply:
Discussion continued in main thread